[Feature] Adding metadata caching capabilities to the jdbc catalog

WholeWorld-Timothy opened the pull request:
Why I'm doing:
The metadata information of the jdbc catalog is not cached, and extensive access to metadata can cause significant pressure on the jdbc database. Increasing the metadata caching capability can solve this problem.
What I'm doing:
Adding metadata caching capabilities to the jdbc catalog. Increase the switch and expiration time of parameter control cache.

Configure FE dynamic parameters for default value:

##### jdbc_meta_default_cache_enable

- **Unit**: -
- **Default**: FALSE
- **Description**: The default value for whether the JDBC Catalog metadata cache is enabled. When set to True, newly created JDBC Catalogs will default to metadata caching enabled.

##### jdbc_meta_default_cache_expire_sec

- **Unit**: Second
- **Default**: 600
- **Description**: The default expiration time for the JDBC Catalog metadata cache. When jdbc_meta_default_cache_enable is set to true, newly created JDBC Catalogs will default to setting the expiration time of the metadata cache.

When jdbc_meta_default_cache_enable parameter is true, the newly created jdbc catalog will automatically enable caching.

Operations like this can enable caching:

```
CREATE EXTERNAL CATALOG `cache_catalog`
PROPERTIES (
......
"jdbc_meta_cache_enable" = "false",
"jdbc_meta_cache_expire_sec" = "600"
)
```

## What type of PR is this:

- [ ] BugFix
- [X] Feature
- [ ] Enhancement
- [ ] Refactor
- [ ] UT
- [x] Doc
- [ ] Tool

Does this PR entail a change in behavior?

- [ ] Yes, this PR will result in a change in behavior.
- [X] No, this PR will not result in a change in behavior.

If yes, please specify the type of change:

- [ ] Interface/UI changes: syntax, type conversion, expression evaluation, display information
- [ ] Parameter changes: default values, similar parameters but with different default values
- [ ] Policy changes: use new policy to replace old one, functionality automatically enabled
- [ ] Feature removed
- [ ] Miscellaneous: upgrade & downgrade compatibility, etc.

## Checklist:

- [X] I have added test cases for my bug fix or my new feature
- [X] This pr needs user documentation (for new or modified features or behaviors)
  - [X] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:

- [X] I have checked the version labels which the pr will be auto-backported to the target branch
  - [ ] 3.2
  - [ ] 3.1
  - [ ] 3.0
  - [ ] 2.5

WholeWorld-Timothy committed:
Add metadata cache for jdbc catalog

Signed-off-by: 张田 <zhang_tian@inspur.com>

WholeWorld-Timothy committed:
Add test cases for metadata cache with jdbc catalog

Signed-off-by: 张田 <zhang_tian@inspur.com>

WholeWorld-Timothy committed:
fixed code and docs style issue for metadata cache with jdbc catalog. fixed bug for metadata cache with jdbc catalog.

Signed-off-by: 张田 <zhang_tian@inspur.com>

WholeWorld-Timothy committed:
Optimize test cases for metadata cache with jdbc catalog.

Signed-off-by: 张田 <zhang_tian@inspur.com>

WholeWorld-Timothy committed:
fixed issue 34320

Signed-off-by: 张田 <zhang_tian@inspur.com>

WholeWorld-Timothy committed:
fixed issue 34320

Signed-off-by: 张田 <zhang_tian@inspur.com>

WholeWorld-Timothy committed:
merge pr fixed-34320

Signed-off-by: 张田 <zhang_tian@inspur.com>

WholeWorld-Timothy committed:
merge pr fixed-34320

Signed-off-by: 张田 <zhang_tian@inspur.com>

zombee0 commented on fe/fe-core/src/main/java/com/starrocks/common/Config.java:
```diff
@@ -2656,4 +2656,10 @@ public class Config extends ConfigBase {
+    public static boolean jdbc_meta_cache_enable = false;
+
+    @ConfField(mutable = true)
+    public static long jdbc_meta_cache_expire_sec = 600L;
```
Jdbc metadata usually is stable, so we can enlarge expire time. And other metadata cache' configuration usually use session variable.

zombee0 commented on fe/fe-core/src/main/java/com/starrocks/common/Config.java:
```diff
@@ -2656,4 +2656,10 @@ public class Config extends ConfigBase {
+    public static boolean jdbc_meta_cache_enable = false;
+
+    @ConfField(mutable = true)
+    public static long jdbc_meta_cache_expire_sec = 600L;
```
Jdbc metadata usually is stable, so we can enlarge expire time. And other metadata cache' configuration usually use session variable.

WholeWorld-Timothy replied:
600 seconds is an empirical value. From a practical perspective, the metadata of MySQL may not be as stable. This is an empirical value we obtained from other projects that falls between performance and cache time. I think we can give it a try first. If there are good suggestions from customers, we can make the necessary changes.

zombee0 commented on fe/fe-core/src/main/java/com/starrocks/connector/jdbc/JDBCMetadata.java:
```diff
@@ -15,44 +15,41 @@
+    private JDBCMetaResolver metaResolver;
+
+    private final @NonNull JDBCAsyncCache<ImmutableMap<String, String>, List<String>> catalogCache
+            = new JDBCAsyncCache<>();
```
For final, How about that there are two jdbc catalog?

WholeWorld-Timothy replied:
The structure is as follows: each data source has only one jdbc catalog implementation class, and each implementation class contains some cache classes. My understanding is that there is only one jdbc catalog, but have I not understood your question yet?

zombee0 commented on fe/fe-core/src/main/java/com/starrocks/connector/jdbc/JDBCMetadata.java:
```diff
@@ -76,11 +74,9 @@ public Connection getConnection() throws SQLException {
-        }
+        ImmutableMap<String, String> metaInfo =
+                ImmutableMap.of("func", "listDbNames");
+        return catalogCache.get(metaInfo, k -> metaResolver.refreshCacheForListDbNames());
```
as async cache, if it doesn't hit on cache, it will return null? if you need wait the result why not use sync cache?

WholeWorld-Timothy replied:
The asynchrony here does not mean that the thread has returned, but rather that the thread is suspended and waiting for the result to return. For high concurrency databases, this is a good way.

WholeWorld-Timothy replied:
For more information, please refer to : https://zhuanlan.zhihu.com/p/611934066

zombee0 commented on fe/fe-core/src/main/java/com/starrocks/connector/jdbc/JDBCMetadata.java:
```diff
@@ -76,11 +74,9 @@ public Connection getConnection() throws SQLException {
-        }
+        ImmutableMap<String, String> metaInfo =
+                ImmutableMap.of("func", "listDbNames");
+        return catalogCache.get(metaInfo, k -> metaResolver.refreshCacheForListDbNames());
```
as async cache, if it doesn't hit on cache, it will return null? if you need wait the result why not use sync cache?

WholeWorld-Timothy committed:
Optimize code writing style

Signed-off-by: 张田 <zhang_tian@inspur.com>

WholeWorld-Timothy commented on fe/fe-core/src/main/java/com/starrocks/connector/jdbc/JDBCSchemaResolver.java:
```diff
@@ -96,4 +98,14 @@ public List<Column> convertToSRTable(ResultSet columnSet) throws SQLException {
     }
+
+    public boolean checkAndSetSupportPartitionInformation(Connection connection) {
+        return false;
```
Method return for test cases

WholeWorld-Timothy committed:
fixed some bugs for jdbc catalog cache

Signed-off-by: 张田 <zhang_tian@inspur.com>

WholeWorld-Timothy committed:
fixed some bugs for jdbc catalog cache test unit

Signed-off-by: 张田 <zhang_tian@inspur.com>

WholeWorld-Timothy committed:
fixed some bugs for jdbc catalog cache test unit

Signed-off-by: 张田 <zhang_tian@inspur.com>

WholeWorld-Timothy commented on fe/fe-core/src/main/java/com/starrocks/connector/jdbc/JDBCAsyncCache.java:
```diff
@@ -0,0 +1,54 @@
+    }
+
+    public @NonNull V get(@NonNull K key, @NonNull Function<K, V> function) {
+        if (Config.jdbc_meta_cache_enable) {
```
It is necessary to configure cache parameters according to the catalog, but no method to modify the catalog properties has been found in the current code. The dynamic switches of the cache and the modification of the cache time are crucial dynamic parameters

WholeWorld-Timothy replied:
I suggest starting a new PR that includes use catalog properties and refresh_external_table

Youngwb replied:
use config will change all jdbc catalogs cache if you modify the jdbc_meta_default_cache_enable or jdbc_meta_default_cache_expire_sec. it's better to use catalog properties

WholeWorld-Timothy replied:
This version has been changed to use catalog properties

WholeWorld-Timothy committed:
Add JDBCCacheKey and cancel the use of map

Signed-off-by: 张田 <zhang_tian@inspur.com>

stephen-shelby commented on fe/fe-core/src/main/java/com/starrocks/connector/jdbc/JDBCMetaResolver.java:
```diff
@@ -0,0 +1,183 @@
+        try (Connection connection = getConnection()) {
+            return Lists.newArrayList(schemaResolver.listSchemas(connection));
+        } catch (SQLException e) {
+            throw new StarRocksConnectorException(e.getMessage());
```
it's best to print error stack to locate issue.

WholeWorld-Timothy replied:
replace with 'throw new StarRocksConnectorException("refresh db names cache for JDBC catalog fail!", e);'

WholeWorld-Timothy committed:
Implementation of adding method REFRESH INTERNATIONAL TABLE.
Delete two classes: JDBCMetaResolver.java, JDBCTableCacheKey.java.
Add test cases.

Signed-off-by: 张田 <zhang_tian@inspur.com>

WholeWorld-Timothy commented:
Implementation of adding method REFRESH INTERNATIONAL TABLE.
Delete two classes: JDBCMetaResolver.java, JDBCTableCacheKey.java.
Add some test cases.

WholeWorld-Timothy committed:
Add cache configuration in the properties of the catalog.

Signed-off-by: 张田 <zhang_tian@inspur.com>

WholeWorld-Timothy committed:
Change the implementation method of cache key

Signed-off-by: 张田 <zhang_tian@inspur.com>

Youngwb commented on fe/fe-core/src/main/java/com/starrocks/connector/jdbc/JDBCAsyncCache.java:
```diff
@@ -0,0 +1,93 @@
+
+    public @NonNull V get(@NonNull K key, @NonNull Function<K, V> function) {
+        if (this.enableCache) {
+            return this.asyncCache.get(key, function).join();
```
if you use join, what's the difference between async cache and sync cache

WholeWorld-Timothy replied:
Misunderstanding of the use of async cache, changed to sync cache

zombee0 commented on fe/fe-core/src/main/java/com/starrocks/connector/jdbc/JDBCMetadata.java:
```diff
@@ -68,27 +73,36 @@ public JDBCMetadata(Map<String, String> properties, String catalogName) {
             return Lists.newArrayList(schemaResolver.listSchemas(connection));
         } catch (SQLException e) {
-            throw new StarRocksConnectorException(e.getMessage());
+            throw new StarRocksConnectorException("refresh db names cache for JDBC catalog fail!", e);
```
if that user dose not enable cache and throw this exception is puzzling

WholeWorld-Timothy replied:
done

zombee0 commented on fe/fe-core/src/main/java/com/starrocks/connector/jdbc/JDBCMetadata.java:
```diff
@@ -117,89 +131,119 @@ public List<String> listTableNames(String dbName) {
             }
         } catch (SQLException e) {
-            throw new StarRocksConnectorException(e.getMessage());
+            throw new StarRocksConnectorException("refresh table names cache for JDBC catalog fail!", e);
```
ditto

WholeWorld-Timothy replied:
done

zombee0 commented on fe/fe-core/src/main/java/com/starrocks/connector/jdbc/JDBCMetadata.java:
```diff
@@ -117,89 +131,119 @@ public List<String> listTableNames(String dbName) {
+                    try (Connection connection = getConnection()) {
+                        return schemaResolver.listPartitionNames(connection, databaseName, tableName);
+                    } catch (SQLException e) {
+                        throw new StarRocksConnectorException("refresh partition names cache for JDBC catalog fail!",
```
ditto

WholeWorld-Timothy replied:
done

WholeWorld-Timothy committed:
Change asynchronous cache to synchronous cache

Signed-off-by: 张田 <zhang_tian@inspur.com>
